Make sure inactive request generator is stopped before spawning new one in Pub / Sub consumer.

dhermes opened the pull request:
~~This makes the `request_queue` used by `Consumer._request_generator_thread()` local to the caller. This way on recovery a new queue can be used for the new thread.~~

lukesneeringer commented on pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py:
```diff
@@ -244,7 +252,19 @@ def _blocking_consume(self):
                 recover = self._policy.on_exception(exc)
-                if not recover:
+                if recover:
+                    with threading.Lock():
```
I am fairly certain that `with threading.Lock():` used in this way does not work, because it is always using a different lock. You need to save a single lock object to the consumer instance so you are calling the `__enter__` method of the same lock.

Basically, `self._recovery_lock = threading.Lock()` in the `__init__.py` file, and then `with self._recovery_lock:` here.

dhermes replied:
What is the issue with always using a different lock?

lukesneeringer replied:
It will never actually lock anything.

dhermes replied:
BLERG

lukesneeringer replied:
Longer explanation: `threading.Lock()` creates a single, discrete lock. To illustrate, imagine a world where we have two separate locks that lock two separate things:

To illustrate, consider a kitchen with two cooks. Both cooks want to make scrambled eggs on the stove top, and they want to put bread in the toaster. Since there is only one stove top and only one toaster, the cooks can not use them at the same time, but it would be fine for one cook to use the stove top while the other one uses the toaster.

These are different lock instances:

```python
stove_lock = threading.Lock()
toaster_lock = threading.Lock()

with stove_lock:
    scramble_eggs()

with toaster_lock:
    toast_bread()
```

When you call `with toaster_lock:`, if another thread is currently executing code within a separate `with toaster_lock:`, then your thread will block until the other thread calls `toaster_lock.__exit__`. However, it does not matter if another thread is currently using `stove_lock`.

But when you just do `with threading.Lock():` then you are creating a _different_ lock instance every time, and it is not actually possible for anything else to be checking against the same lock. The other threads are just blindly using their own, distinct locks and not checking in with yours.

lukesneeringer commented on pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py:
```diff
@@ -60,12 +60,13 @@ def __init__(self):
         return needle in self._helper_threads
 
-    def start(self, name, queue, target):
+    def start(self, name, queue_put, target):
```
What is the rationale behind this change?

dhermes replied:
See where I am sending in `Consumer.send_request()` rather than passing in `Consumer._request_queue`. The problem is that if you send in a queue and then swap it off of the `Consumer`, then the helper thread won't be able to talk to the active request generator.

lukesneeringer replied:
Understood.

dhermes replied:
@lukesneeringer The rationale has change slightly, since I no longer swap out the request queue. However, due to the `_put_lock` in `send_request()`, it still makes sense to avoid passing the queue itself.

dhermes commented on pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py:
```diff
@@ -244,7 +252,19 @@ def _blocking_consume(self):
+                    with threading.Lock():
+                        # Must lock so that ``self.send_request()`` cannot add
+                        # more items to the queue while it is being replaced.
+                        previous_queue = self._request_queue
```
@lukesneeringer After chatting with @jonparrott it seems there is another option here.

If we could guarantee that `self._request_queue.empty()` is `True`, then we could just send `STOP` into the queue and wait until the `STOP` gets consumed, then we don't have to swap the queue.

(As you can see from the coverage failure, the queue will always be empty.)

lukesneeringer replied:
That makes sense to me.

dhermes replied:
#4504 is an implementation of that other option

dhermes commented:
@jonparrott @lukesneeringer PTAL

dhermes commented on pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py:
```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
+            try:
+                request_generator.close()
+            except ValueError as exc:
+                if exc.args != ('generator already executing',):
```
This is not so great to have this be "stringly-typed", I could just assume that a `ValueError` means it's already executing.

lukesneeringer replied:
I trust your judgment on this.

dhermes commented on pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py:
```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
+        """
+        with self._put_lock:
+            try:
+                request_generator.close()
```
There **might** is a seemingly impossible-to-trigger corner case where the generator is **still running** for the few nanoseconds **after** the blocking `.get()` returns. In that case, we'd either need to

- Somehow force the gRPC thread that is consuming the request generator to knock it off
- Just try to `.close()` one more time if we detect the queue isn't empty

----

FWIW I don't think it's worth accounting for that impossible-to-trigger corner case.

jonparrott commented on pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py:
```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
+                # queue **was** and remains empty.
+                self._request_queue.put(_helper_threads.STOP)
+                # Wait for the request generator to ``.get()`` the ``STOP``.
+                while not self._request_queue.empty():
```
if we use `task_done` we can use [queue.join](https://docs.python.org/2/library/queue.html#Queue.Queue.join).

jonparrott commented on pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py:
```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
+            except ValueError as exc:
+                if exc.args != ('generator already executing',):
+                    raise
+                if not self._request_queue.empty():
```
I thought we discussed that it's perfectly valid for the queue to have some pending items in it?

dhermes commented on pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py:
```diff
@@ -218,6 +223,54 @@ def _request_generator_thread(self):
+                    'Waiting for active request generator to receive STOP')
+                while not self._request_queue.empty():
+                    pass
+                request_generator.close()
```
Luckily I re-factored the unit tests, because it made me realize that this could fail, because the generator still has to do:

```python
if request == _helper_threads.STOP:
    logger.debug()
    break
```

after the blocking `get()` completes, so this call to `close()` might happen **before** the generator gets those instructions.

I'm inclined to say we should either:

- Drop this statement completely (we are "safe" since we know that `STOP` has been pulled off the queue)
- Do a `while True` loop that tries to `close()` and `break`-s if it succeeds
- Do a `while request_generator.gi_running: pass` loop to block until it stops running and then call `close()`

jonparrott replied:
I am in favor of dropping `close()` as we know that as long as `STOP` (or *any* item in the queue) has been consumed post-RPC-cancellation that gRPC will exit the thread.

dhermes replied:
Me too. Jinx.

dhermes commented:
@lukesneeringer @jonparrott PTAL. Phew I think it's finally merge-able.

jonparrott commented on pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py:
```diff
@@ -218,6 +223,71 @@ def _request_generator_thread(self):
+                        'request queue is not empty.')
+                    return False
+                # If we **cannot** close the request generator,
+                # then there is no blocking get on the queue. Since
```
Do you mean that there *is* a blocking get?

dhermes replied:
Yes, thanks.

dhermes replied:
Updated

jonparrott commented on pubsub/tests/unit/pubsub_v1/subscriber/test_consumer.py:
```diff
@@ -142,6 +140,106 @@ def test_start_consuming():
+        yield value
+
+
+def test_stop_request_generator_not_running():
```
I would love some comments explaining the scenarios that each test exercise.

dhermes commented:
I added explanations for each unit test @jonparrott. Also updated one of them to use a real generator instead of a `mock`. Will merge once CI is green.

This code is no fun.
